modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/JavaExtensionModelParser.java

```diff
@@ -147,12 +155,64 @@ private void parseStructure(ExtensionElement extensionElement) {
 
   private Set<String> parseSupportedJavaVersions(ExtensionElement extensionElement) {
     return extensionElement.getValueFromAnnotation(JavaVersionSupport.class)
-        .map(a -> a.getEnumArrayValue(JavaVersionSupport::value).stream()
+        .map(a -> a
+            .getEnumArrayValue(javaVersionSupport -> getJavaVersionSupport(javaVersionSupport, extensionElement))
+            .stream()
             .map(JavaVersion::version)
             .collect(toCollection(() -> (Set<String>) new LinkedHashSet<String>())))
         .orElse(emptySet());
   }
 
+  private JavaVersion[] getJavaVersionSupport(JavaVersionSupport javaVersionSupport, ExtensionElement extensionElement) {
+    try {
+      return javaVersionSupport.value();
+    } catch (EnumConstantNotPresentException e) {
+      // An enum value not present in the Mule container version of `JavaVersion` is in use, we need to introspect the annotation
+    }
+
+    try {
+      Class<?> annotatedClass = extensionElement
+          .getDeclaringClass()
+          .orElseThrow(() -> new RuntimeException(format("Expected extension element '%s' to have a declaring class",
+                                                         extensionElement)));
+      ClassReader reader = new ClassReader(currentThread().getContextClassLoader()
+          .getResourceAsStream(annotatedClass.getName().replace(".", "/") + ".class"));
+
+      List<JavaVersion> javaVersions = new ArrayList<>();
+      reader.accept(new ClassVisitor(ASM9) {
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
+          if (descriptor.endsWith(JavaVersionSupport.class.getSimpleName() + ";")) {
+            return new AnnotationVisitor(ASM9) {
+
+              @Override
+              public AnnotationVisitor visitArray(String name) {
+                return new AnnotationVisitor(ASM9) {
+
+                  @Override
+                  public void visitEnum(String name, String desc, String value) {
+                    try {
+                      javaVersions.add(valueOf(value));
+                    } catch (Exception e) {
+                      LOGGER.debug("Found unknown value '{}' in JavaVersionSupport annotation", value, e);
+                    }
+                  }
+                };
+              }
+            };
+          }
+
+          return super.visitAnnotation(descriptor, visible);
+        }
+      }, 0);
+
+      return javaVersions.toArray(new JavaVersion[javaVersions.size()]);
+    } catch (Exception e) {
+      throw new RuntimeException("Unexpected error while parsing JavaVersionSupport values", e);
```

Comment:
rethrow the original EnumConstantNotPresentException with this `e` as suppressed.

Reply:
Done.